File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        if(heights.length==0){
            return 0;
        }
        if(heights.length==1){
            return heights[0];
        }

        int[] left = nse_left(heights);
        int[] right = nse_right(heights);

        int ans = 0;
        for(int i=0; i<heights.length; i++){
            int res = right[i]-left[i]-1;
            ans = (ans>res*heights[i]) ? ans : res*heights[i];
        }

        return ans;
    }

    public int[] nse_left(int[] arr){
        int[] res = new int[arr.length];
        res[0] = -1;
        Stack<Integer> st = new Stack<>();
        st.push(0);
        for(int i=1; i<arr.length; i++){
            if(arr[st.peek()]<arr[i]){
                res[i] = st.peek();
                st.push(i);
            }
            else{
                while(!(st.isEmpty()) && arr[st.peek()]>=arr[i]){
                    st.pop();
                }
                if(st.isEmpty()){
                    res[i] = -1;
                    st.push(i);
                }
                else{
                    res[i] = st.peek();
                    st.push(i);
                }
            }
        }

        return res;
    }

    public int[] nse_right(int[] arr){
        int[] res = new int[arr.length];
        res[arr.length-1] = arr.length;
        Stack<Integer> st = new Stack<>();
        st.push(arr.length-1);
        for(int i = arr.length-2; i>=0; i--){
            if(arr[st.peek()]<arr[i]){
                res[i] = st.peek();
                st.push(i);
            }
            else{
                while(!(st.isEmpty()) && arr[st.peek()]>=arr[i]){
                    st.pop();
                }
                if(st.isEmpty()){
                    st.push(i);
                    res[i] = arr.length;
                }
                else{
                    res[i] = st.peek();
                    st.push(i);
                }
            }
        }

        return res;
    }
}

File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        if(len(heights)==0):
            return 0
        if(len(heights)==1):
            return heights[0]
        
        left = self.nse_left(heights)
        right = self.nse_right(heights)

        ans = 0
        for i in range(len(heights)):
            val = right[i]-left[i]-1
            ans = max(ans, val*heights[i])

        return ans

    def nse_left(self, arr):
        res = [0]*len(arr)
        res[0] = -1
        st = [0]
        for i in range(1, len(arr)):
            if(arr[st[-1]]<arr[i]):
                res[i] = st[-1]
                st.append(i)
            else:
                while(not(len(st)==0) and arr[st[-1]]>=arr[i]):
                    st.pop()
                if(len(st)==0):
                    res[i] = -1
                    st.append(i)
                else:
                    res[i] = st[-1]
                    st.append(i)
        
        return res

    def nse_right(self, arr):
        res = [0]*len(arr)
        res[len(arr)-1] = len(arr)
        st = [len(arr)-1]
        for i in range(len(arr)-2, -1, -1):
            if(arr[st[-1]]<arr[i]):
                res[i] = st[-1]
                st.append(i)
            else:
                while(not(len(st)==0) and arr[st[-1]]>=arr[i]):
                    st.pop()
                if(len(st)==0):
                    res[i] = len(arr)
                    st.append(i)
                else:
                    res[i] = st[-1]
                    st.append(i)
        
        return res